File: 6th_train/3I.py
n = int(input())
a, b = map(int, input().split())
roads = list((list(), list(), list(), list()))
main_roads = [False]*4
main_roads[a % 4] = True
main_roads[b % 4] = True

for i in range(n):
    road, time = map(int, input().split());
    roads[road % 4].append([time, i])

times = list()
for i in range(4):
    roads[i].sort(key=lambda rover: rover[0])
    if roads[i]:
        times.append(roads[i][0][0])

time = min(times)
passed_time = [0]*n

while(not all((not road) for road in roads)):
    current_state = [list()]*4
    passed = [False]*4
    for i in range(4):
        if (roads[i]):
            if (roads[i][0][0] <= time):
                current_state[i] = roads[i][0]
                passed[i] = True

    for i in range(4):
        if(passed[i]):
            if (not main_roads[i]):
                if ( (main_roads[i-1] and current_state[i-1]) or (main_roads[(i+1)%4] and current_state[(i+1) % 4]) or (main_roads[(i+2)%4] and current_state[(i+2) % 4])):
                    passed[i] = False

            if (main_roads[i-1] == main_roads[i] and current_state[i-1]):
                    passed[i] = False
   
    for i in range(4):
        if (passed[i]):
           passed_time[current_state[i][1]] = time
           roads[i].pop(0)
    
    time += 1

for i in passed_time:
    print(i)

File: 6th_train/3I.cpp
#include "iostream"
#include "stdlib.h"
#include "algorithm"
#include "queue"

using namespace std;

struct rover
{
  int n;
  int road;
  int time;
};

bool comp(rover a1, rover a2)
{
  return(a1.time < a2.time);
}

int main()
{
  int N, a, b, *order;
  bool main[4] = {0};
  
  rover *list;
  cin >> N >> a >> b;
  main[a%4] = true;
  main[b%4] = true;
  list = (rover*)calloc(N, sizeof(rover));
  order = (int *)calloc(N, sizeof(int));
  for (int i = 0; i < N; i++)
  {
    cin >> list[i].road >> list[i].time;
    list[i].n = i;
  }
  sort(list, list+N, comp);
  
  queue<rover> road[4];
  bool goaway[4];
  int passed = 0, getto = 0, time = 0;
  while(passed < N)
  {
    time++;
    while (getto < N)
    {
      if(list[getto].time <= time)
      {
        road[list[getto].road % 4].push(list[getto]);
        getto++;
      }
      else break;
    }
    for (int i = 4; i < 8; i ++)
    {
      goaway[i%4] = false;
      if (!road[i % 4].empty())
      {
        if (main[i%4])
        {
          if (!main[(i-1) % 4] or road[(i-1) % 4].empty())
          {
            goaway[i%4] = true;
          }
        }
        else if (road[(i-1) % 4].empty() and (!main[(i+1) % 4] or road[(i+1) % 4].empty()))
        {
          goaway[i%4] = true;
        }
      }
    }
    for (int i = 4; i < 8; i ++)
      if (goaway[i%4])
      {
        passed++;
        order[(road[i%4].front()).n] = time;
        road[i%4].pop();
      }
  }
  
  for (int i = 0; i < N; i ++)
    cout << order[i] << '\n';
  
}
